### ISSUE TITLE
Add initial documentation

### ISSUE THREAD

**Issue by piersy** [clearmatics/autonity#490]
# Problem

Following on from #461 this is a ticket that covers adding initial documentation for autonity. The plan was to have myself and @maxcroasdale develop documentation for a few of the required documentation headings, so that we have enough documentation to set the style/tone/approach such that subsequent contributors have enough examples of existing documentation that they can easily match the existing style/tone/approach.

As such we will add documentation for a subset of the agreed upon documentation headings for the next release.
```
How To -> Run a node (covers build, install, config, running)
How To -> Run a network (covers configuring operator, participants, validators)
How To -> Access JSON-RPC API
```

# Acceptance Criteria

Documentation for the aforementioned headings should be available in the `docs` folder.

**Comment by JekaMas** [comment]
Shall it be in this repo or in a separate one?

**Comment by piersy** [comment]
@JekaMas The docs are going to be held within this repo, under the docs dir.

**Comment by JekaMas** [comment]
I'm thinking about the license. Wouldn't it be more suitable for us to create a new repo with something non-gpl and keep docs there?

@yazzaoui

**Comment by yazzaoui** [comment]
I think we can always specify in the README which part of the repo is under our own copyright?

**Comment by JekaMas** [comment]
We can't. There's the license for the whole repo.

**Comment by piersy** [comment]
@JekaMas Could you explain why you think the licence could pose a problem with the docs?

**Comment by JekaMas** [comment]
We'll not be able to move docs to a private repo. For me, we should use private repos everywhere where it's possible just to reduce our future work with migration to a new license.

**Comment by piersy** [comment]
@JekaMas @yazzaoui Can we just license the docs folder differently, we could put an MIT licence in it?

**Comment by JekaMas** [comment]
As far as I know, it doesn't work that way. You cant put different license into the folder. Moreover, the docs folder doesn't have there own license by the time and we can't just introduce a new one different from gnuv3. We even can't use anything that is in docs in a private directory.

**Comment by JekaMas** [comment]
I'd suggest to be on a safe side and create a new repo. We don't have enough experience in licensing so it's easy to do something wrong.

**Comment by saronimo** [comment]
I could see this being an issue if we can't use any of the content in a private file as we wouldn't be able to use to core how-to docs. That said, the list of topics outline by @piersy in #461 should be the basis for the oss docs anyway.

@piersy what are the implications of having the docs in a separate repo?

**Comment by piersy** [comment]
@saronimo It makes it much harder to keep the code and documentation in sync.

We would not be able to change some code and update the corresponding docs in an atomic manner. Instead we would need to open a separate pr for documentation in a docs repo, meaning that there will always be some time delay between doc being updated and code being updated. Either the code is going to get merged first or the doc will, either way that leaves them out of sync for some time period.

If you consider this happening across multiple prs (we currently have 14 open prs) and features, it could turn out that the documentation rarely represent's the state of the code, which could render the doc not very helpful most of the time.

By contrast having the doc and code in the same repo means that users of the repo can rely on their checkouts always having doc that correctly represents the current state of the code. 

If we used a separate repo for docs, when doing releases we would also need to allow some "stabilisation" time where we ensure all relevant doc/code has been merged for the current release.

So unless there is a compelling reason otherwise I would like to keep the doc in the same repo as the code.

Since this is OSS documentation, it seems right that it should live with the source under the same terms. But I'm pretty sure we could license this under a different license if we wished, because it is a separate program from the autonity system and so would be considered ["mere aggregation"](https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.en.html#MereAggregation)

**Comment by JekaMas** [comment]
I think we can do it. We can use https://www.atlassian.com/git/tutorials/git-submodule and keep the documentation in a separate repo and have PRs that are atomic: we could work in Autonity repo but include into PRs changes from docs submodule - https://codeyarns.com/2016/02/10/how-to-work-with-git-submodules/

**Comment by saronimo** [comment]
Thanks @piersy I get the issue, that does seem cumbersome.

So if I understand you correctly. we could combine two repos, one with the oss docs covered under the LGPM v3.0 and combine it with docs originating in an other repo, covered by a different license, but both displayed side-by-side?

**Comment by piersy** [comment]
@JekaMas I think practically using submodules to maintain atomic commits will be quite hard.
 
PRs in autonity would need to wait for the corresponding doc PR to be merged before they could correctly reference the commit hash. But once a doc PR has been merged, we would then need to wait for its corresponding autonity PR to be merged before any newer autonity PRs with doc changes  could be merged, because otherwise those newer PRs would be adding doc for the feature that would not have been merged yet, thus resulting in out of sync doc and code.

Once a doc PR had been merged any change to the PR in autonity could require that the doc be updated (so we'd need to open a new pr for the doc and wait for that to get merged again) and then remember to update the commit hash of the doc in the main PR. Whilst all that is happening we would not be able to merge newer PRs with doc, for the reason outlined above.

I've worked in a project where we used submodules a lot, and people would forget to update the commit hashes, this resulted in deploying branches or worse having submodule commit hashes that don't point to anything in the remote.

So if at all possible I'd like to not go down this route and instead, keep doc in the same repo as the code.

I'm not sure what you mean by
>*The thing we still need deal with: we already have docs on unsuitable license. And we can't just put license file to the folder.

Are you saying that by adding the docs folder initially without a licence it now falls under the GPL licence, so we cant just change it to be under a different licence?

**Comment by piersy** [comment]
@saronimo

>So if I understand you correctly. we could combine two repos, one with the oss docs covered under the LGPM v3.0 and combine it with docs originating in an other repo, covered by a different license, but both displayed side-by-side?

I'm not fully clear on what you mean here.

What I was suggesting is that we could use a different licence for the docs directory in the autonity project. The autonity already has a mix of licences that are used depending on which bits of code you are looking at, so I'm just suggesting we could do the same.

I'm not clear on why we would want to "combine two repos" or do anything else with this doc other than publish it in this repo.

If we can establish that this documentation just covers the autonity system and it just lives here in the autonity repo, then I don't see any issue with it being covered by the licences that are already present.

**Comment by JekaMas** [comment]
Not really. With git submodules you can
1. Fetch master repo
2. Fetch submodule
3. Work in submodule like it's a folder inside master repo. Do any changes you need.
4. Create PRs for master repo and submodule at the same time.

And we already use submodules in Autonity for tests 'tests/testdata'. It's not hard or complicated at all.

**Comment by piersy** [comment]
@JekaMas Submodules work well when the submodule doesn't change very often.

In the case of the testdata submodule it has been updated in 14 commits out of 1850 since it was added as a submodule.

When the submodules are changing quite often and especially if you have multiple open PRs on a submodule then things become a bit trckier, as I outlined [here](https://github.com/clearmatics/autonity/issues/490#issuecomment-606623557).

At least I think we can definitely agree that keeping stuff in sync that is in the same repo, is simpler than using submodules.

**Comment by JekaMas** [comment]
We don't change API or contracts often. Maybe once a month or even rarely.

**Comment by JekaMas** [comment]
Tests submodule has >2400 commits https://github.com/ethereum/tests?files=1
I think it's quite a lot.

**Comment by JekaMas** [comment]
For me user and business experience are more priority than development. And we should do our best for them. It's not a big deal to support one repo more.

**Comment by piersy** [comment]
@JekaMas If user and business experience includes having documentation that matches the code then I think having the doc and code versioned together in a single repo provide a better user and business experience.

**Comment by JekaMas** [comment]
Does it a big deal to keep docs in sync if we're a small team and you an only person who is on docs now?

**Comment by saronimo** [comment]
So at this point it seems to make most sense to keep these in the same repo for ease of keeping the docs in synch with the code. I believe this has the following benefits from a usability perspective 1) it will allow @maxcroasdale to keep track of delta in the docs by existing github issues as a source (don't close until you've updated the docs) and 2) for any additional users that come to the repo and want to contribute there will be synchronised documentation readily available.

I appreciate your concerns around the licensing @JekaMas but given all the code is covered by the same license I'm not sure that it is more pertinent than a accurate documentation of the code

**Comment by piersy** [comment]
@JekaMas It seems that we are ok with keeping the docs under the LGPL licence so we can keep them in the same repository as the code, do you have any other concerns?

**Comment by JekaMas** [comment]
For me, we cannot decide things that we are not expects. I'd prefer keep this things in a new repo.

**Comment by piersy** [comment]
Just noting here that, given the business team is happy for the docs to be licensed along with the code, we have gone ahead and merged #491.

**Comment by piersy** [comment]
Closing this issue because it has been broken down in to subtasks.
